mm2src/mm2_main/src/lp_swap/taker_restart.rs

```diff
@@ -0,0 +1,264 @@
+            tx: &taker_payment,
+            search_from_block: taker_coin_start_block,
+            swap_contract_address: &taker_coin_swap_contract_address,
+            swap_unique_data: &unique_data,
+            watcher_reward,
+        })
+        .await
+}
+
+pub async fn add_taker_payment_spent_event(
+    swap: &TakerSwap,
+    saved: &mut TakerSavedSwap,
+    taker_payment_spend_tx: &TransactionEnum,
+) -> Result<(), String> {
+    let secret_hash = swap.r().secret_hash.0.clone();
+    let watcher_reward = swap.r().watcher_reward;
+
+    let tx_hash = taker_payment_spend_tx.tx_hash();
+    info!("Taker payment spend tx {:02x}", tx_hash);
+    let tx_ident = TransactionIdentifier {
+        tx_hex: Bytes::from(taker_payment_spend_tx.tx_hex()),
+        tx_hash,
+    };
+    let secret = match swap
+        .taker_coin
+        .extract_secret(&secret_hash, &tx_ident.tx_hex, watcher_reward)
+        .await
+    {
+        Ok(bytes) => H256::from(bytes.as_slice()),
+        Err(_) => {
+            return ERR!("Could not extract secret from taker payment spend transaction");
+        },
+    };
+
+    let event = TakerSwapEvent::TakerPaymentSpent(TakerPaymentSpentData {
+        transaction: tx_ident,
+        secret,
+    });
+    let to_save = TakerSavedEvent {
+        timestamp: now_ms(),
+        event,
+    };
+    saved.events.push(to_save);
+    Ok(())
+}
+
+pub async fn add_taker_payment_refunded_by_watcher_event(
+    ctx: &MmArc,
+    swap: &TakerSwap,
+    mut saved: TakerSavedSwap,
+    taker_payment_refund_tx: TransactionEnum,
+) -> Result<TakerSwapCommand, String> {
+    let other_maker_coin_htlc_pub = swap.r().other_maker_coin_htlc_pub;
+    let taker_coin_swap_contract_address = swap.r().data.taker_coin_swap_contract_address.clone();
+    let taker_payment_lock = match std::env::var("USE_TEST_LOCKTIME") {
+        Ok(_) => swap.r().data.started_at as u32,
+        Err(_) => swap.r().data.taker_payment_lock as u32,
+    };
+    let secret_hash = swap.r().secret_hash.0.clone();
+
+    let validate_input = ValidateWatcherSpendInput {
+        payment_tx: taker_payment_refund_tx.tx_hex(),
+        maker_pub: other_maker_coin_htlc_pub.to_vec(),
+        swap_contract_address: taker_coin_swap_contract_address,
+        time_lock: taker_payment_lock,
+        secret_hash: secret_hash.clone(),
+        amount: swap.taker_amount.to_decimal(),
+        watcher_reward: None,
+        spend_type: WatcherSpendType::TakerPaymentRefund,
+    };
+
+    swap.taker_coin
+        .taker_validates_payment_spend_or_refund(validate_input)
+        .compat()
+        .await
+        .map_err(|e| e.to_string())?;
+
+    let tx_hash = taker_payment_refund_tx.tx_hash();
+    info!("Taker refund tx hash {:02x}", tx_hash);
+    let tx_ident = TransactionIdentifier {
+        tx_hex: Bytes::from(taker_payment_refund_tx.tx_hex()),
+        tx_hash,
+    };
+
+    let event = TakerSwapEvent::TakerPaymentRefundedByWatcher(Some(tx_ident));
```

Comment:
This should be included in `TAKER_ERROR_EVENTS`
https://github.com/KomodoPlatform/komodo-defi-framework/blob/1c1d93e8788234ba0f4427b18b82808a69b5a558/mm2src/mm2_main/src/lp_swap/taker_swap.rs#L76-L92